File: Packages/com.gamelovers.mcp-unity/Server~/src/tools/getPlayModeStatusTool.ts
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { McpUnity } from '../unity/mcpUnity.js';
import { McpUnityError, ErrorType } from '../utils/errors.js';
import * as z from 'zod';
import { Logger } from '../utils/logger.js';
import { CallToolResult } from '@modelcontextprotocol/sdk/types.js';

const toolName = 'get_play_mode_status';
const toolDescription = 'Gets Unity play mode status (isPlaying, isPaused).';

const paramsSchema = z.object({});

export function registerGetPlayModeStatusTool(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  logger.info(`Registering tool: ${toolName}`);

  server.tool(
    toolName,
    toolDescription,
    paramsSchema.shape,
    async (params: any) => {
      try {
        logger.info(`Executing tool: ${toolName}`, params);
        const result = await toolHandler(mcpUnity, params);
        logger.info(`Tool execution successful: ${toolName}`);
        return result;
      } catch (error) {
        logger.error(`Tool execution failed: ${toolName}`, error);
        throw error;
      }
    }
  );
}

async function toolHandler(mcpUnity: McpUnity, params: any): Promise<CallToolResult> {
  const response = await mcpUnity.sendRequest({
    method: toolName,
    params
  });

  if (!response.success) {
    throw new McpUnityError(
      ErrorType.TOOL_EXECUTION,
      response.message || 'Failed to get play mode status'
    );
  }

  const statusText = response.isPlaying
    ? (response.isPaused ? 'Play mode (paused)' : 'Play mode')
    : 'Edit mode';

  return {
    content: [
      {
        type: response.type as 'text',
        text: statusText
      }
    ],
    data: {
      isPlaying: response.isPlaying,
      isPaused: response.isPaused
    }
  };
}

File: Packages/com.gamelovers.mcp-unity/Editor/Tools/GetPlayModeStatusTool.cs
using System;
using Newtonsoft.Json.Linq;
using UnityEditor;
using McpUnity.Unity;
using McpUnity.Utils;

namespace McpUnity.Tools
{
    /// <summary>
    /// Tool for getting Unity play mode status
    /// </summary>
    public class GetPlayModeStatusTool : McpToolBase
    {
        public GetPlayModeStatusTool()
        {
            Name = "get_play_mode_status";
            Description = "Gets Unity play mode status (isPlaying, isPaused).";
        }

        public override JObject Execute(JObject parameters)
        {
            try
            {
                bool isPlaying = EditorApplication.isPlaying;
                bool isPaused = EditorApplication.isPaused;

                var result = new JObject
                {
                    ["success"] = true,
                    ["type"] = "text",
                    ["message"] = isPlaying ? (isPaused ? "Play mode (paused)" : "Play mode") : "Edit mode",
                    ["isPlaying"] = isPlaying,
                    ["isPaused"] = isPaused
                };

                McpLogger.LogInfo($"Play mode status requested: isPlaying={isPlaying}, isPaused={isPaused}");

                return result;
            }
            catch (Exception ex)
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    $"Error getting play mode status: {ex.Message}",
                    "play_mode_error"
                );
            }
        }
    }
}
